inference-engine/src/multi_device/multi_device_exec_network.cpp

```diff
@@ -159,109 +159,181 @@ MultiDeviceExecutableNetwork::MultiDeviceExecutableNetwork(const std::string&
+    auto& device = context.deviceInfo.deviceName;
+    auto& deviceConfig = context.deviceInfo.config;
+    auto& deviceList = context.metaDevices;
+    bool curDevIsCPU = (device.find("CPU") != std::string::npos);
+    try {
+        if (!modelPath.empty()) {
+            context.executableNetwork = _core->LoadNetwork(modelPath, device, deviceConfig);
+        } else {
+            context.executableNetwork = _core->LoadNetwork(network, device, deviceConfig);
+        }
+        context.isLoadSuccess = true;
+    } catch (const std::exception& e) {
+        context.errMessage += device + ":" + e.what();
+        context.isLoadSuccess = false;
+    }
 
-            if (device.find("CPU") == std::string::npos) {
-                _alreadyActualNetwork = true;
-                _acceleratorPromise.set_value(executableNetwork);
-            } else {
-                _cpuPromise.set_value(executableNetwork);
-            }
-        });
+    if (context.isLoadSuccess || curDevIsCPU) {
+        return;
     }
 
-    WaitFirstNetworkReady();
+    // remove the current device from deviceList
+    auto eraseDevice = std::find_if(deviceList.begin(), deviceList.end(),
+            [device](DeviceInformation& d){
+            return d.deviceName == device;
+            });
+    deviceList.erase(eraseDevice);
+
+    if (deviceList.empty()) {
+        return;
+    }
+
+    // select next candidate device
+    try {
+        context.deviceInfo = _multiPlugin->SelectDevice(deviceList, context.networkPrecision);
+    }
+    catch (const std::exception& e) {
+        return;
+    }
+
+    // if selec device is CPU, do not need to load CPU again, context[CPU] must have loaded CPU
+    curDevIsCPU = (context.deviceInfo.deviceName.find("CPU") != std::string::npos);
+    if (curDevIsCPU) {
+        return;
+    }
+
+    // try to load this candidate device
+    TryToLoadNetWork(context, modelPath, network);
 }
 
 void MultiDeviceExecutableNetwork::WaitFirstNetworkReady() {
-    if (_alreadyActualNetwork) {
-        return;
+    if (_firstLoadFuture.valid()) {
+        // wait for the first loading finished
+        _firstLoadFuture.wait();
     }
-    if (_cpuFuture.valid() && _acceleratorFuture.valid()) {
-        try {
-            _networkFirstReady = _cpuFuture.get();
-        } catch (const std::exception& e) {
-            printf("Warning: load network to CPU failed: %s\n", e.what());
-            _networkActualNeeded = _acceleratorFuture.get();
+
+    // check if there is any device that have loaded network successfully
+    for (int i = CONTEXTNUM - 1; i >= 0; i--) {
+        if (_loadContext[i].isEnabled && _loadContext[i].isAlready) {
+            return;
+        }
+    }
+
+    // the first loading is failed, wait for another loading
+    for (int i = CONTEXTNUM - 1; i >= 0; i--) {
+        if (_loadContext[i].isEnabled) {
+            _loadContext[i].future.wait();
```

Comment:
not sure why this loop cannot be combined with the below loop... when the wait completed, we are already on the right index, so only need to check isAlready[i]

Reply:
fix it